ccl/streamingccl/streamproducer: TestReplicationStreamTenant failed

ccl/streamingccl/streamproducer.TestReplicationStreamTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186481&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186481&tab=artifacts#/) on master @ [e1068d77afbd39b162978281c9da7cbea49c1c3a](https://github.com/cockroachdb/cockroach/commits/e1068d77afbd39b162978281c9da7cbea49c1c3a):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:43 +0x164
  github.com/cockroachdb/cockroach/pkg/sql.startExec()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:519 +0x180
  github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:146 +0x124
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:158 +0x14b
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:68 +0x97
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:87 +0x58
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedStatsCollectorImpl).Init()
      <autogenerated>:1 +0x4b
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:235 +0x68
  github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x7b
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:234 +0x70
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:275 +0x330
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259 +0x328
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:563 +0xf94
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1401 +0x249
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1408 +0xceb
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1085 +0x12c4
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:722 +0x342a
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:135 +0x93c
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:222 +0x1dc
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1820 +0xa1c
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1822 +0x6f8
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676 +0x5fe
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:681 +0x14b
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:669 +0x405
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestReplicationStreamTenant/stream-tenant-with-cursor (0.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicationStreamTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestReplicationStreamTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4198664&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4198664&tab=artifacts#/) on master @ [1c4b0ddb3c29305db08f9ae848a6c0d0b190f620](https://github.com/cockroachdb/cockroach/commits/1c4b0ddb3c29305db08f9ae848a6c0d0b190f620):


```
=== RUN   TestReplicationStreamTenant
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicationStreamTenant243999275
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestReplicationStreamTenant
    replication_stream_test.go:275: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicationStreamTenant243999275
    testing.go:1152: race detected during execution of test
--- FAIL: TestReplicationStreamTenant (3.39s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicationStreamTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/util/tracing/crdbspan.go
+++ b/pkg/util/tracing/crdbspan.go
@@ -85,7 +85,12 @@ type crdbSpanMu struct {
 	// Finish()).
 	parent spanRef
 
+	// finished is set if finish() was called.
 	finished bool
+	// finishing is set while finish() is in the process of running. finish() has
+	// two separate critical sections, and this finishing field is used to detect
+	// when other calls have been interleaved between them.
+	finishing bool
 
 	// duration is initialized to -1 and set on Finish().
 	duration time.Duration
@@ -231,14 +236,18 @@ func (s *crdbSpan) finish() bool {
 		parent = s.mu.parent.move()
 		hasParent = !parent.empty()
 		if hasParent {
+			s.mu.finishing = true
 			s.mu.Unlock()
 		}
 	}
 
 	// Operate on the parent outside the child (our current receiver) lock.
-	// childFinished() might call back into the child and acquire the child's
-	// lock.
+	// childFinished() might call back into the child (`s`) and acquire the
+	// child's lock.
 	if hasParent {
+		// It's possible to race with parent.Finish(); if we lose the race, the
+		// parent will not have any record of this child. childFinished() deals with
+		// that possibility.
 		parent.Span.i.crdb.childFinished(s)
 		parent.release()
 	}
@@ -250,6 +259,7 @@ func (s *crdbSpan) finish() bool {
 		// Re-acquire the lock if we dropped it above.
 		if hasParent {
 			s.mu.Lock()
+			s.mu.finishing = false
 		}
 
 		// If the span is not part of the registry now, it never will be. So, we'll
@@ -264,12 +274,14 @@ func (s *crdbSpan) finish() bool {
 		//
 		// We also shallow-copy the children for operating on them outside the lock.
 		children = make([]spanRef, len(s.mu.openChildren))
-		for i, c := range s.mu.openChildren {
+		for i := range s.mu.openChildren {
+			c := &s.mu.openChildren[i]
 			c.parentFinished()
 			// Move ownership of the child reference, and also nil out the pointer to
 			// the child, making it available for GC.
 			children[i] = c.spanRef.move()
 		}
+		s.mu.openChildren = nil // The children were moved away.
 		s.mu.Unlock()
 	}
 
@@ -722,13 +734,15 @@ func (s *crdbSpan) addChildLocked(child *Span, collectChildRec bool) bool {
 // with its recording.
 //
 // child is the child span that just finished.
+//
+// This method can be called while the parent (s) is still alive, is in the
+// process of finishing (i.e. s.finish() is in the middle section where it has
+// dropped the lock), or has completed finishing. In the first two cases, it is
+// expected that child is part of s.mu.openChildren. In the last case, this will
+// be a no-op.
 func (s *crdbSpan) childFinished(child *crdbSpan) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-
-	if s.mu.finished {
-		return
-	}
 
 	var childIdx int
 	found := false
@@ -741,10 +755,35 @@ func (s *crdbSpan) childFinished(child *crdbSpan) {
 		}
 	}
 	if !found {
-		// The child is expected to be found. In particular, this parent span is not
-		// supposed to be reused while children are holding a reference to it
-		// (courtesy of the parent's reference counter).
-		panic("child not present in parent")
+		// We haven't found the child, so the child must be calling into the parent
+		// (s) after s.finish() completed. This can happen if finishing the child
+		// races with finishing the parent (if s.finish() would have ran to
+		// completion before c.finish() was called, c wouldn't have called
+		// s.childFinished(c) because it would no longer have had a parent).
+		if !s.mu.finished {
+			panic("unexpectedly failed to find child of non-finished parent")
+		}
+		// Since s has been finished, there's nothing more to do.
+		return
+	}
+
+	// Sanity check the situations when this method is called after s.finish() was
+	// called. Running this after s.finish() was called indicates that s.finish()
+	// and child.finish() were called concurrently - if s.finish() would have
+	// finished before child.finish() was called, the child would not have had a
+	// parent to call childFinished() on. Here, we assert that we understand the
+	// races that can happen, as this concurrency is subtle.
+	// - s.childFinished(c) can be called after s.finish() completed, in which
+	//   case `found` will be false above, and we won't get here.
+	// - s.childFinished(c) can also be called after s.finish() started, but
+	//   before s.finish() completed. s.finish() has two critical sections, and
+	//   drops its lock in between them.
+	// This is the only case where the child is found, and we explicitly assert
+	// that here.
+	if s.mu.finished {
+		if !s.mu.finishing {
+			panic("child unexpectedly calling into finished parent")
+		}
 	}
 
 	collectChildRec := s.mu.openChildren[childIdx].collectRecording

--- a/pkg/util/tracing/tracer.go
+++ b/pkg/util/tracing/tracer.go
@@ -486,6 +486,10 @@ type TracerTestingKnobs struct {
 	// MaintainAllocationCounters, if set,  configures the Tracer to maintain
 	// counters about span creation. See Tracer.GetStatsAndReset().
 	MaintainAllocationCounters bool
+	// ReleaseSpanToPool, if set, if called just before a span is put in the
+	// sync.Pool for reuse. If the hook returns false, the span will not be put in
+	// the pool.
+	ReleaseSpanToPool func(*Span) bool
 }
 
 // Redactable returns true if the tracer is configured to emit
@@ -943,7 +947,13 @@ func (t *Tracer) releaseSpanToPool(sp *Span) {
 	h.childrenAlloc = [4]childRef{}
 	h.structuredEventsAlloc = [3]interface{}{}
 
-	t.spanPool.Put(h)
+	release := true
+	if fn := t.testing.ReleaseSpanToPool; fn != nil {
+		release = fn(sp)
+	}
+	if release {
+		t.spanPool.Put(h)
+	}
 }
 
 // StartSpanCtx starts a Span and returns it alongside a wrapping Context
@@ -959,6 +969,9 @@ func (t *Tracer) StartSpanCtx(
 	// `opts` on the heap here.
 	var opts spanOptions
 	for _, o := range os {
+		if o == nil {
+			continue
+		}
 		opts = o.apply(opts)
 	}
 

--- a/pkg/util/tracing/tracer_test.go
+++ b/pkg/util/tracing/tracer_test.go
@@ -12,6 +12,8 @@ package tracing
 
 import (
 	"context"
+	"fmt"
+	"math/rand"
 	"sort"
 	"sync"
 	"testing"
@@ -756,4 +758,58 @@ func TestSpanReuse(t *testing.T) {
 		}
 	}
 	t.Fatal("spans do not seem to be reused reliably")
+}
+
+// Test that parents and children can finish in any order (even concurrently)
+// and that the recording is always sane in such cases.
+func TestSpanFinishRaces(t *testing.T) {
+	ctx := context.Background()
+	tr := NewTracerWithOpt(ctx,
+		WithSpanReusePercent(100),
+		WithTracingMode(TracingModeActiveSpansRegistry),
+		WithTestingKnobs(TracerTestingKnobs{
+			ReleaseSpanToPool: func(sp *Span) bool {
+				// Asynchronously overwrite sp, to catch races.
+				go func() {
+					*sp = Span{}
+				}()
+				// Tell the tracer to not pool this span. We've hijacked it above.
+				return false
+			}}),
+	)
+
+	const numSpans = 4
+
+	for ti := 0; ti < 1000; ti++ {
+		sps := make([]*Span, numSpans)
+		for i := 0; i < numSpans; i++ {
+			var opt SpanOption
+			if i > 0 {
+				opt = WithParent(sps[i-1])
+			}
+			sps[i] = tr.StartSpan(fmt.Sprint(i), WithRecording(RecordingVerbose), opt)
+			sps[i].Recordf("msg %d", i)
+		}
+
+		finishOrder := rand.Perm(numSpans)
+		var rec Recording
+		g := sync.WaitGroup{}
+		g.Add(len(finishOrder))
+		for _, idx := range finishOrder {
+			go func(idx int) {
+				if idx != 0 {
+					sps[idx].Finish()
+				} else {
+					rec = sps[0].FinishAndGetRecording(RecordingVerbose)
+				}
+				g.Done()
+			}(idx)
+		}
+		g.Wait()
+		require.Len(t, rec, numSpans)
+		for i, s := range rec {
+			require.Len(t, s.Logs, 1)
+			require.Equal(t, fmt.Sprintf("msg %d", i), s.Logs[0].Message.StripMarkers())
+		}
+	}
 }

